flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataStreamProcTimeAggregateWindowFunction.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.runtime.aggregate
+
+import java.lang.Iterable
+
+import org.apache.flink.api.java.tuple.Tuple
+import org.apache.flink.types.Row
+import org.apache.flink.configuration.Configuration
+import org.apache.flink.streaming.api.functions.windowing.RichWindowFunction
+import org.apache.flink.streaming.api.windowing.windows.Window
+import org.apache.flink.util.Collector
+import org.apache.flink.table.functions.AggregateFunction
+import org.apache.flink.table.functions.Accumulator
+
+/**
+  * Computes the final aggregate value from incrementally computed aggreagtes.
+  *
+  * @param numGroupingKey The number of grouping keys.
+  * @param numAggregates The number of aggregates.
+  * @param finalRowArity The arity of the final output row.
+  */
+class DataStreamIncrementalAggregateWindowFunction[W <: Window](
+     private val aggregates: Array[AggregateFunction[_]],
+     private val aggFields: Array[Int],
+     private val forwardedFieldCount: Int)
+  extends RichWindowFunction[Row, Row, Tuple, W] {
+
+private var output: Row = _
+private var accumulators: Row= _
+
+  override def open(parameters: Configuration): Unit = {
+     output = new Row(forwardedFieldCount + aggregates.length)
+     accumulators = new Row(aggregates.length)
+     var i = 0
```

Comment:
We can do that in `open()` because it is just called once. However, in the hot path we should use `while` loops because Scala `for` loops have significant overhead.